### ISSUE TITLE
about the playready drm...

### ISSUE THREAD

**Issue by hereisclient** [google/ExoPlayer#8131]
### [REQUIRED] Exoplayer couldn't play  vod content with playready.

First time, i doubted the security level issue.
but the playready version was 2.0 in the MPD file
So this case is not security issue.
https://docs.microsoft.com/en-us/playready/overview/security-level
That's why i help exo-team.

I've got a test hbbtv app from the customer company in germany.
This app has a video content protected by  the playready drm.
i tested this video using my test exoplayer app.
But it was failed.
So i requested to the customer whether the drm content is right or not.
the customer sent me one video clip playing drm content.
Unfortunately , some of device played the same contents.

So i just want know why the exoplay can't play that drm content.

### [REQUIRED] Reproduction steps
just add content url on playready list in media_exolist.json file
and install exoplay and lunch an app and play added vod content.

### [REQUIRED] Link to test content
I've already sent it exo team 

### [REQUIRED] A full bug report captured from the device
I've already sent it exo team 

### [REQUIRED] Version of ExoPlayer being used
version : 2.9.4

### [REQUIRED] Device(s) and version(s) of Android being used
android device : shaomi mi box 3
android version : 9

It might also be helpful if you could try and play the PlayReady sample in the ExoPlayer demo app on the affected device, to see if that plays successfully. That will help you determine whether it's an issue with the content or an issue with the device.
**-> every playready contents can play successfully except this issue case.
also,   this time is not my first time playing the vod contents encryted by the playready.**

As a final comment, is there a reason the client is not using Widevine? It's quite likely that they'll need to use Widevine for targeting some other devices anyway, and support for it is generally more robust across Android than support for PlayReady.
**-> this test vod content is for hbbtv app. and  the standard of hbbtv 2.0 supports the playready instead of the widevine. that's why uses the playready.**

**Comment by ojw28** [comment]
It's not possible to say from the information you've provided. You'll need to provide the more complete information requested in the [bug template](https://github.com/google/ExoPlayer/issues/new?assignees=&labels=bug%2C+needs+triage&template=bug.md&title=), rather than the question template you filled in. If you do this, please copy/paste the template here rather than opening a new issue.

It might also be helpful if you could try and play the PlayReady sample in the ExoPlayer demo app on the affected device, to see if that plays successfully. That will help you determine whether it's an issue with the content or an issue with the device.

As a final comment, is there a reason the client is not using Widevine? It's quite likely that they'll need to use Widevine for targeting some other devices anyway, and support for it is generally more robust across Android than support for PlayReady.

**Comment by hereisclient** [comment]
i updated my question  by your request.
please check it again.
Thank you so much.

**Comment by ojw28** [comment]
I'm not sure whether you sent the correct stream URL, since:

1. There's some logging present in the log file you sent that's content dependent, and is different to the logging I see when playing the provided stream (e.g., the duration is slightly different).
1. The provided stream works fine for me on MiBox3 running Android 9.
1. The provided stream is not geo-blocked as described.

Please double check. If it is the correct stream, then I'm unable to reproduce this locally, which is going to make it difficult for us to investigate further. I also tried on nVidia Shield, and the stream player correctly on this device too.

**Comment by hereisclient** [comment]
I'm not sure whether you sent the correct stream URL, since:

1.There's some logging present in the log file you sent that's content dependent, and is different to the logging I see when playing the provided stream (e.g., the duration is slightly different).
    **I tested it again after exoplayer uninstallation and installation.
    After that the error log was changed .
    I emailed with the new error log.
    i don't know this what stream is .  but i guess if it is live stream, the duration will not fixed .**  

2.The provided stream works fine for me on MiBox3 running Android 9.
    **i test it with 2.9.4 exoplayer version and 2.12.1 but the result was the same.
    All my modification was just adding test content in media_exolist.json file. 
    Maybe, is there any difference comparing my code and your code?
    and You can see the screen and hear the sound instead of black screen with mute. Is that right?**

3.The provided stream is not geo-blocked as described.
    **i'm not sure that, i didn't hear any information from the customer,
    but by now at least playready drm contents have been blocked by region, So i just assumed. if you played it in your area, that 
    doesn't have geo-block.**

i added my explanation.

thanks for your help.

**Comment by ojw28** [comment]
I tried playing the stream again, and am still able to play it without any issue. I'm not making any code changes to test this. I'm just checking out the r2.9.4 release tag and adding the sample to the `media.exolist.json` file.

In your latest log, it appears that the DRM license server is an [500 Internal Server Error](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500). It should be possible for whoever controls the license server to debug why this is happening. It does not seem to be an ExoPlayer issue.

**Comment by hereisclient** [comment]
I couldn't follow this situation. OMG...

if you don't mind sharing your log and media.exolist.json file and your test enviroment  , Please share it for me.
1.log file on xiomi mibox3 with exoplayer version r2.9.4   .
2. media.exolist.json file
3. Timezone on mibox3.
4. if other special setting value exist, please let me know.  

i'm suffering  because of this issue for almost 2 weeks.
even not me, other chinese developers  also are facing the same situation.

So i'm not sure , but  i might be missing the something point now.
In addition ,  i have to solve this issue, because in future , this part job will be assigned to me.

i don't know what is the wrong at least now.
if you share your infomations, i'll try this as copy strategy.

i need one more the drm content to check .
about this, i emailed you with another content.

i can't thank you enough.

**Comment by ojw28** [comment]
Regarding the streams that you previously provided, which work fine for me:

* As per my previous response, if the license server is responding with a HTTP 500 Internal Server Error, you should look at debugging the issue on the server side to see what's wrong.
* The `media.exolist.json` file was modified by exactly copying what you provided in your emails. I don't think the timezone is related, and there are no other special settings that I'm aware of.

Regarding the newly provided stream:

* I can reproduce the "Failed to get key request" error that you see when playing this stream. This issue was fixed in `2.10.0`. I think the fix was probably [this commit](https://github.com/google/ExoPlayer/commit/282cf303a4ad321fa97e73faa8a133c130c8dd95#diff-94b2168a173755f90142381a28157103c9357bfccbab8d21423cd65d6c14fc30), but I'm not certain. I am able to play the content successfully using 2.10.0.

Regarding 2.12.0 and 2.12.1, unfortunately, there is a separate issue affecting playback of DRM protected content in the demo app where `drm_license_url` is omitted from the `media.exolist.json`. You can work around this by adding
```
"drm_license_url": "https://ignored.com"
```
into your `media.exolist.json` entries. The actual URL is never used, but specifying one avoids the issue. We will fix this in 2.12.2.

So in summary:

* The "Failed to get key request" error was fixed in the 2.10.0 release. Using this release or later allows your stream that sees this error to play successfully. If using 2.12.1 or 2.12.2, you additionally need to use the workaround described above.
* The HTTP 500 Internal Server Error is something you'll need to debug on the license server side.

**Comment by hereisclient** [comment]
The "Failed to get key request" error was fixed in the 2.10.0 release. Using this release or later allows your stream that sees this error to play successfully. If using 2.12.1 or 2.12.2, you additionally need to use the workaround described above.
**i tested it following the your guideline, but it brought the http 500 internal error
   In result, two vod contents occured the same error.**


The HTTP 500 Internal Server Error is something you'll need to debug on the license server side.
**it's not easy to ask that broadcast company about the http 500 error.
the reason is like the below.**
**1. your device and other device could play it successfully under the same test environment.
(just i guess there might be little difference. but your answer was clear...)**
**2. it takes really long time to get reply about my ask from the involving company. i guess it will be over 2 months)
but i think there is no option to choose for me except 2.**
**i really appreciate your helps and supports.**

**Comment by ojw28** [comment]
Closing because I think the question is answered, to the extent to which it's possible to answer it. Any difficulty you might have getting support from a broadcast company is unfortunately not something we can assist with.
